ws/public/main.js

```diff
@@ -0,0 +1,145 @@
+        }
+
+        drawLine(data.from, data.to, data.color)
+    }
+
+    const drawLine = (from, to, color) => {
+        context.beginPath()
+        context.moveTo(from.x, from.y)
+        context.lineTo(to.x, to.y)
+        context.strokeStyle = color
+        context.lineWidth = 2
+        context.stroke()
+        context.closePath()
+    }
+
+    const draw = (emit = false) => {
+        drawLine(state.from, state.to, state.color)
+
+        if (!emit) {
+            return
+        }
+
+        const canvasWidth = canvas.width
+        const canvasHeight = canvas.height
+
+        socket.send(JSON.stringify({
+            name: 'drawing',
+            data: {
+                from: {
+                    x: state.from.x / canvasWidth,
+                    y: state.from.y / canvasHeight
+                },
+                to: {
+                    x: state.to.x / canvasWidth,
+                    y: state.to.y / canvasHeight
+                },
+                color: state.color
+            }
+        }))
+    }
+
+    const onMouseDown = (event) => {
+        state.drawing = true
+
+        state.from.x = event.clientX || event.touches[0].clientX
+        state.from.y = event.clientY || event.touches[0].clientY
+    }
+
+    const onMouseMove = (event) => {
+        if (!state.drawing) {
+            return
+        }
+
+        state.to.x = event.clientX || event.touches[0].clientX
+        state.to.y = event.clientY || event.touches[0].clientY
+
+        draw(true)
+
+        state.from.x = event.clientX || event.touches[0].clientX
+        state.from.y = event.clientY || event.touches[0].clientY
+    }
+
+    const onMouseUp = (event) => {
+        if (!state.drawing) {
+            return
+        }
+
+        state.drawing = false
+        state.to.x = event.clientX || event.touches[0].clientX
+        state.to.y = event.clientY || event.touches[0].clientY
+
+        draw(true);
+    }
+
+    // Resize canvas
+    const setCanvas = () => {
+        canvas.width = window.innerWidth
+        canvas.height = window.innerHeight
+    }
+
+    canvas.addEventListener('mousedown', onMouseDown, false)
+    canvas.addEventListener('mouseup', onMouseUp, false)
+    canvas.addEventListener('mouseout', onMouseUp, false)
+    canvas.addEventListener('mousemove', onMouseMove, false)
+
+    //Touch support for mobile devices
+    canvas.addEventListener('touchstart', onMouseDown, false)
+    canvas.addEventListener('touchend', onMouseUp, false)
+    canvas.addEventListener('touchcancel', onMouseUp, false)
+    canvas.addEventListener('touchmove', onMouseMove, false)
+
+
+    document.getElementsByClassName('colors')[0].addEventListener('click', (event) => {
+        if (event.target.className.includes('colors')) {
+            return
+        }
+
+        document.getElementsByClassName('active')[0].classList.remove('active')
+
+        state.color = event.target.getAttribute('data-color');
+        event.target.classList.add('active')
```

Comment:
I find this logic could be written in a simpler way. e.g. you're adding an event listener on the parent but then you're returning early if it's the parent that was clicked. also getting a list of elements to always get the first one doesn't feel right and suggest that's a better way to write this

Reply:
Sorry but i don't understand the comment.
I am adding the event on the parent and checking if the click was on one of the colors. The list of active should consist one single element being active at any time.